p751toolbox/curve.go

```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
 	return xQ
 }
 
+// Given xP = x(P), xQ = x(Q), and xPmQ = x(P-Q), compute xPaddQ = x(P+Q) and  x2P = x(2P).
+func xDblAdd(curve *CachedCurveParameters, xP, xQ, xPmQ *ProjectivePoint) (x2P, xPaddQ ProjectivePoint) {
+	var A, AA, B, BB, C, D, E, DA, CB, t0, t1 ExtensionFieldElement
+	x1, z1 := &xPmQ.X, &xPmQ.Z
+	x2, z2 := &xP.X, &xP.Z
+	x3, z3 := &xQ.X, &xQ.Z
+
+	A.Add(x2, z2) // A = x2+z2
+	B.Sub(x2, z2) // B = x2-z2
+	C.Add(x3, z3) // C = x3+z3
+	D.Sub(x3, z3) // D = x3-z3
+
+	AA.Square(&A) // AA = A^2
+	BB.Square(&B) // BB = B^2
+
+	E.Sub(&AA, &BB)        // E = AA-BB
+	BB.Mul(&BB, &curve.C4) // BB = (4C)*BB
+	DA.Mul(&D, &A)         // DA = D*A
+	CB.Mul(&C, &B)         // CB = C*B
+
+	t1.Add(&DA, &CB) // t1 = DA+CB
+	t0.Sub(&DA, &CB) // t0 = DA-CB
+	t1.Square(&t1)   // t1 = t1^2
+	t0.Square(&t0)   // t0 = t0^2
+
+	xPaddQ.X.Mul(z1, &t1) // z5 = z1*t1
+	xPaddQ.Z.Mul(x1, &t0) // x5 = x1*t0
+
+	x2P.X.Mul(&AA, &BB)           // x4 = AA*(4C)*BB
+	x2P.Z.Mul(&curve.Aplus2C, &E) // z4 = (A+2C)*E
+	x2P.Z.Add(&x2P.Z, &BB)        // z4 = (4C)*BB+(A+2C)*E
+	x2P.Z.Mul(&x2P.Z, &E)         // z4 = E*((4C)*BB+(A+2C)*E)
+	return
+}
+
+// Given xP = x(P), xQ = x(Q), and xPmQ = x(P-Q), compute xPaddQ = x(P+Q) and  x2P = x(2P).
+// Assumes that the Z-xoordinate of PmQ is equal to 1.
+func xDblAdd_primefield(aPlus2Over4 *PrimeFieldElement, xP, xQ, xPmQ *ProjectivePrimeFieldPoint) (x2P, xPaddQ ProjectivePrimeFieldPoint) {
+	var A, AA, B, BB, C, D, E, DA, CB, t0, t1 PrimeFieldElement
+	x1 := &xPmQ.X
+	x2, z2 := &xP.X, &xP.Z
+	x3, z3 := &xQ.X, &xQ.Z
+
+	A.Add(x2, z2) // A = x2+z2
+	B.Sub(x2, z2) // B = x2-z2
+	C.Add(x3, z3) // C = x3+z3
+	D.Sub(x3, z3) // D = x3-z3
+
+	AA.Square(&A) // AA = A^2
+	BB.Square(&B) // BB = B^2
+
+	E.Sub(&AA, &BB) // E = AA-BB
+	DA.Mul(&D, &A)  // DA = D*A
+	CB.Mul(&C, &B)  // CB = C*B
+
+	t1.Add(&DA, &CB) // t1 = DA+CB
+	t0.Sub(&DA, &CB) // t0 = DA-CB
```

Comment:
For example reuse DA, instead t0, or even right into z5 (xPaddQ.X?) to avoid copying.